interfaces/builtin/mount_control.go

```diff
@@ -0,0 +1,337 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/utils"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/systemd"
+)
+
+const mountControlSummary = `allows creating transient and persistent mounts`
+
+const mountControlBaseDeclarationPlugs = `
+  mount-control:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+const mountControlBaseDeclarationSlots = `
+  mount-control:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-connection: true
+`
+
+const mountAttrNotADictionary = `mount-control "mount" attribute is not a dictionary`
+const mountControlConnectedPlugSecComp = `
+# Description: Allow mount and umount syscall access.
+mount
+umount
+umount2
+`
+
+var allowedMountOptions = []string{
+	"async",
+	"atime",
+	"bind",
+	"diratime",
+	"dirsync",
+	"iversion",
+	"lazytime",
+	"nofail",
+	"noiversion",
+	"nomand",
+	"noatime",
+	"nodev",
+	"nodiratime",
+	"noexec",
+	"nolazytime",
+	"norelatime",
+	"nosuid",
+	"nostrictatime",
+	"nouser",
+	"relatime",
+	"strictatime",
+	"sync",
+	"ro",
+	"rw",
+}
+
+// mountControlInterface allows creating transient and persistent mounts
+type mountControlInterface struct {
+	commonInterface
+}
+
+// TODO: review the regexp
+// We prohibit expressions containing a comma, as that is used by AppArmor to
+// separate rules; otherwise, one could specify a path like
+//   $SNAP/foo, /** rw
+// and get read-write access to the whole filesystem.
+var whereRegexp = regexp.MustCompile(`^(/media|\$SNAP|\$SNAP_COMMON|\$SNAP_DATA)/[^,]*$`)
```

Comment:
have we thought about also allowing $SNAP_USER_DATA or $SNAP_USER_COMMON ? Probably not super useful since the mount operation must be performed by root, so those just resolve to `/root/snap/<snap-name>/{common,current}/` but maybe there is a use case?

Reply:
The spec document only mentions (in the text or in the examples) only `/media`, `$SNAP_COMMON` and `$SNAP_DATA`. I'll leave it to @pedronis to decide if adding more makes sense.

I now realized that I added `$SNAP` to the list, but maybe it should be removed? This would also address the other  remark of yours, where you point out the weirdness of letting a snap shadow its `meta/` directory. Actually, I'll remove it for now :-) I'll add it back if *somuelse* (shorthand for "Samuele or someone else" ;-) ) asks for it.